mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,929 @@
+        Self::set_range_type(addr, len, MttMemoryType::NonConfidential)
+    }
+
+    /// Returns where region of length `len` starting with `addr` is confidential in
+    /// the Memory Tracking Table (MTT).
+    /// 1: `len` must be a multiple of a 4K-page size. The region cannot exceed the
+    /// maximum physical page address (currently defined as 46-bits of PA).
+    pub fn is_confidential(addr: SupervisorPageAddr, len: usize) -> Result<bool> {
+        Self::is_range_of_type(addr, len, MttMemoryType::Confidential)
+    }
+
+    /// Returns where region of length `len` starting with `addr` is non-confidential in
+    /// the Memory Tracking Table (MTT).
+    /// 1: `len` must be a multiple of a 4K-page size. The region cannot exceed the
+    /// maximum physical page address (currently defined as 46-bits of PA).
+    pub fn is_non_confidential(addr: SupervisorPageAddr, len: usize) -> Result<bool> {
+        Self::is_range_of_type(addr, len, MttMemoryType::NonConfidential)
+    }
+}
+
+trait MemoryType {}
+struct Conf;
+struct NonConf;
+impl MemoryType for Conf {}
+impl MemoryType for NonConf {}
+
+// Enumeration of defined L2 entry types.
+// The MTT (Memory Tracking Table) L2 is a 8MB physically contiguous range of memory allocated
+// by trusted platform-FW. When enabled in hardware, it partitions memory into non-confidential
+// and confidential regions. The enumerations below correspond to the values that can be encoded
+// in the 64-bit entries in the table. The entries are encoded as follows:
+// |--------------------------------------------------
+// |  63:36: Zero | 35:34: Type | 33:0: Info
+// |---------------------------------------------------
+// The 2-bit encoding for Type is as follows:
+// 00b: Non-confidential 1GB range
+// 01b: Confidential 1GB range
+// 10b: A 64-MB range further described by 16K entries in a 4K page (pointed to by Info[33:0])
+// 11b: A 64-MB range composed of 32x2MB sub-ranges, and further described by Info[31:0]
+enum MttL2Entry {
+    // The entire 1GB range containing the address is non-confidential.
+    // 1GB ranges are mapped using 64MB subranges, and by convention, the invariant is
+    // that each of the 16 consecutive entries in the L2 table have the same type.
+    NonConfidential1G(Mtt1GEntry<NonConf>),
+    // The entire 1GB range containing the address is confidential.
+    // 1GB ranges are mapped using 64MB subranges, and by convention, the invariant is
+    // that each of the 16 consecutive entries in the L2 table have the same type.
+    Confidential1G(Mtt1GEntry<Conf>),
+    // The 64-MB range has been partitioned into 2MB regions, and each sub-region can be
+    // confidential or non-confidential
+    Mixed2M(Mtt64MEntry),
+    // The 64-MB range has been partitioned into 16K regions of 4KB size.
+    // Each sub-region can be confidential or non-confidential
+    L1Entry(MttL1Entry),
+}
+use MttL2Entry::*;
+
+struct MttL2Directory;
```

Comment:
The different types here still aren't making a ton of sense to me. See my comment from earlier: https://github.com/rivosinc/salus/pull/206#discussion_r1100668215. Why can't we do something like that?

Reply:
As mentioned before, I have don't have an intuitive feel for what the mapping should look like, and have tried to base if on your suggestion (within the boundaries of what the borrow checker will permit). See if this version is more to your liking.